packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts

```diff
@@ -128,6 +128,36 @@ export default createRule<Options, MessageIds>({
       );
     }
 
+    function isTypeUnchanged(uncast: ts.Type, cast: ts.Type): boolean {
+      if (uncast === cast) {
+        return true;
+      } else if (
+        isTypeFlagSet(uncast, ts.TypeFlags.Undefined) &&
+        isTypeFlagSet(cast, ts.TypeFlags.Undefined) &&
+        tsutils.isCompilerOptionEnabled(
+          compilerOptions,
+          'exactOptionalPropertyTypes',
+        )
+      ) {
+        const uncastParts = tsutils
+          .unionTypeParts(uncast)
+          .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+        const castParts = tsutils
+          .unionTypeParts(cast)
+          .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+        if (uncastParts.length !== castParts.length) {
+          return false;
+        }
+
+        const castPartsSet = new Set(castParts);
+        uncastParts.forEach(part => castPartsSet.delete(part));
+        return castPartsSet.size === 0;
```

Comment:
[Style] A teeny teeny nit if you're interested ✨ 

```suggestion
        const uncastPartsSet = new Set(uncastParts);
        return castParts.some(part => uncastPartsSet.has(part));
```

Reply:
Hi @JoshuaKGoldberg  Thanks! probably meant "every", right? I changed it - 72a9a55

Reply:
...yes 😄 that.

Reply:
I'd suggest `Set.symmetricDifference` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/symmetricDifference), but it's not widely supported :smile: